### PR TITLE
Warning on wrong call order

### PR DESCRIPTION
**Pull request by SkafteNicki**
# Before submitting

- [x] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [x] Did you make sure to update the docs?   
- [x] Did you write any new necessary tests?  

## What does this PR do?
Warn user if `compute` is called before `update` as most metrics are meaningless or will produce errors if their states have not been updated. Will return 0.0 in this case.

## PR review    
Anyone in the community is free to review the PR once the tests have passed.     
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: tests/bases/test_metric.py [hunk 1]
```diff
@@ -259,3 +259,25 @@ def test_device_and_dtype_transfer(tmpdir):
...
+    # make sure everything is fine with forward
+    with pytest.warns(None) as record:
+        val = metric(1)
+    assert not record
+
+    metric.reset()
+
+    with pytest.warns(UserWarning, match=r'The ``compute`` method of metric .*'):
+        val = metric.compute()
+        assert val == 0.0
```
Comment (Borda) [inline tests/bases/test_metric.py:276]:
```suggestion
    assert val == 0.0
```

File: torchmetrics/metric.py [hunk 2]
```diff
@@ -206,6 +208,15 @@ def _wrap_compute(self, compute):
...
         @functools.wraps(compute)
         def wrapped_func(*args, **kwargs):
+            if not self._update_called:
+                rank_zero_warn(
+                    f"The ``compute`` method of metric {self.__class__.__name__}"
+                    " was called before the ``update`` method which may lead to errors,"
+                    " as metric states have not yet been updated. Will return 0.0 instead.",
+                    UserWarning
+                )
+                return 0.0
```
Comment (Borda) [inline torchmetrics/metric.py:218]: shall this be rather NaN?

Comment (SkafteNicki) [reply torchmetrics/metric.py:218]: Fine with me :]
What does lightning do in `self.log` if a `NaN` is passed in?

Comment (Borda) [reply torchmetrics/metric.py:218]: @tchaton ^^

Comment (tchaton) [reply torchmetrics/metric.py:218]: Great question. Looks like Lightning doesn't handle NaN values. It will store it and then produce a NaN value at reduction time.
@SkafteNicki Can you have a look ? I think we might want to skip saving the result object is NaN is logged.

Comment (tchaton) [reply torchmetrics/metric.py:218]: And 0 would actually make the reduction False.

Comment (SkafteNicki) [reply torchmetrics/metric.py:218]: let me try to update this on the lightning side

Comment (maximsch2) [reply torchmetrics/metric.py:218]: I'm not sure if it's worth it to return 0.0 - some metrics are returning tuples/vectors when computed correctly and there will be code to handle it, in that case, there will be a weird exception down the line due to type-mismatch. Might be better to just raise instead.

In terms of type-mismatch, this return 0 will also break type-checking if it was smart enough to understand all the dynamic wrapping happening in the first place :)

Comment (SkafteNicki) [reply torchmetrics/metric.py:218]: @maximsch2 so you propose that we raise an error instead?

**Comment by Borda** [comment]
@SkafteNicki is this ready to go?

**Comment by Borda** [comment]
@maximsch2 mind check comments/questions from @SkafteNicki?

File: torchmetrics/metric.py (hunk 2 above)
Comment (Borda) [reply torchmetrics/metric.py:223]: coming back, how about the NaN instead of 0?

Comment (maximsch2) [reply torchmetrics/metric.py:223]: The question is - do we want the training to proceed and do we expect this to happen deep into the training? 

Given that most of the time we will probably get the error right away (on_step=True in training_step and validation will be caught by sanity checking), I think we should throw an exception here. Alternatively - maybe just warn and do nothing? Let `compute` fail if it does? 

Another aspect to think about - say I have a metric that is stateless and the only thing it does is that in `compute()` it queries RAM utilization or CPU load or something else and returns it. Given that it's going to be rare and probably done by more advanced users, asking them to call update() can be fair game I guess.

Comment (SkafteNicki) [reply torchmetrics/metric.py:223]: then lets settle for raising an warning to the users. Then it depends on the metric what will happen afterwards:
* metrics using list states will fail because `compute` tries to call `torch.cat` on empty list
* metrics of the form `num/denom` will give `nan` as it in most cases it will be `0/0`
* some metrics will behave nicely (like confusionmatrix with no normalization) because the metric is well defined from the initial state

**Review by Borda — APPROVED** [review]
Cool =)

**Comment by maximsch2** [comment]
LGTM with the warning.
